README
megatools - command line client application for Mega
====================================================

Megatools is a collection of programs for accessing Mega service from
a command line of your desktop or server.

Megatools allow you to copy individual files as well as entire directory
trees to and from the cloud. You can also perform streaming downloads for
example to preview videos and audio files, without needing to download
the entire file.

Megatools are robust and optimized for fast operation - as fast as Mega
servers allow. Memory requirements and CPU utilization are kept at minimum.

You can register account using a 'megareg' tool, with the benefit of having
true control of your encryption keys.

Mega website can be found at https://mega.nz.

Megatools can be downloaded at https://megatools.megous.com


Tools
=====

  megareg      Register and verify a new mega account
  megadf       Show your cloud storage space usage/quota
  megals       List all remote files
  megamkdir    Create remote directory
  megarm       Remove remote file or directory
  megaput      Upload individual files
  megaget      Download individual files
  megadl       Download file from a "public" Mega link
               (doesn't require login)
  megacopy     Upload or download a directory tree


All of these tools do:

- Local caching of remote session/filesystem information
  for faster execution. Cache is encrypted with your password
  key.
- Support loading login credentials from a configuration file


Usage
=====

See man pages for how to use individual tools and how to configure them:

  man megatools

Man pages are also available online at:

  https://megatools.megous.com/man/megatools.html


Installation on Windows
=======================

Megatools is available on Windows via Chocolatey thanks to ERap320. See:

  https://chocolatey.org/packages/megatools/

You can contact the package maintainer here:

  https://github.com/megous/megatools/issues/347


Windows Quirks
==============

On Windows, if you see weird characters in your megals output, you'll need
to set correct CHARSET environment variable. For example on Czech Windows
this would mean executing this command in cmd before using the tools:

  set CHARSET=CP852

This is just a cosmetic issue. Internally, megatools always work with UTF-8
file names, and even if the tool's terminal output is corrupted, files names
of downloaded/uploaded files will be correct.


Installation on Mac OS X (Mountain Lion)
========================================

Thanks to Carl Moden, megatools is available in Homebrew (http://brew.sh/).

You can therefore install megatools with:

  brew install megatools


Installation on your favorite GNU/Linux distribution
====================================================

Megatools may already be pre-packaged in the package repository
of your distribution. It is already available at least in:

- Debian - https://packages.debian.org/stretch/megatools
- Ubuntu - https://launchpad.net/ubuntu/+source/megatools
- Fedora - https://admin.fedoraproject.org/pkgdb/package/rpms/megatools/
- Arch Linux (AUR) - https://aur.archlinux.org/packages/megatools/
- Gentoo - https://packages.gentoo.org/packages/net-misc/megatools

Be sure to check your distribution's package repository first.


Installation on FreeBSD
=======================

Megatools is available in ports thanks to Maxim V. Kostikov:

  http://www.freshports.org/net/megatools/


Building megatools from source code
===================================

The official source code tarball is available at:

   http://megatools.megous.com/builds/

You will need to install some dependnencies before you can build megatools from
source code. Package names of these dependencies differ depending on your GNU/Linux
distribution.

You should also check that the code was released by me, by verifying PGP signatures
provided alongside the code.

As these tarballs are released infrequently, you can also build code from the master
branch of megatools git repository. You can get the code using git:

  git clone git://megous.com/megatools

NOTE: Please don't report build issues against code that you downloaded from github.
These reports will be closed without explanation, with reference to this README
file. It is not expected to work on all distributions. Use the official tarball.

The necessary dependencies are: glib2, libcurl, openssl

For building from git repository, you'll also need: asciidoc (or use --disable-docs)

If you don't have those already, some basic development packages are also necessary.
Namely: gcc, make, pkg-config (pkgconf), and when building from git: autoconf,
automake.

On Debian, Ubuntu:

  apt-get -y install build-essential libglib2.0-dev libssl-dev \
    libcurl4-openssl-dev

On Fedora and CentOS:

  yum -y install gcc make glib2-devel libcurl-devel openssl-devel

  If you're building megatools on CentOS or other simialrly old system, you may
  need to pass -std=gnu99 to the C comipler.

On OpenSUSE:

  zypper -n install gcc make glib2-devel libcurl-devel openssl-devel

On Arch Linux:

  pacman -Sy --noconfirm --needed pkgconf gcc make glib2 curl

On Alpine Linux:

  apk add --update build-base libcurl curl-dev asciidoc openssl-dev glib-dev \
    glib libtool automake autoconf


You can build megatools into your HOME directory, so that they'll not pollute
your /usr or /usr/local by using --prefix=$HOME/.local option to configure.

If you're building from git repository, use ./autogen.sh instead of ./configure
and vice versa.

Example build steps:

  git clone git://megous.com/megatools
  cd megatools
  ./autogen.sh --prefix=$HOME/.local --disable-docs
  make -j4
  make install

Now you can run megatools from ~/.local/bin.

  export PATH="$HOME/.local/bin:$PATH"
  megals

If you encounter issues, read the error messages carefully. They may contain
hints on how you can solve the issue yourself.


Third party tools/scripts
=========================

Megatools are meant as a low-level tools that can be used as a base to create
more complicated tools for working with mega.nz. Other people have created
a third party scripts on top of megatools. If you're one of them and want to
have your project listed here, let me know.

Some third party tools can be found at:

  https://amourspirit.github.io/mega_scripts/


Author
======

Megatools were written by Ondřej Jirman <megous@megous.com>, 2014-2018

My PGP key can be found at: https://megous.com/pgp
  (Fingerprint is: 9AB138B20691621CD4CF92026E6426C677CFEFF1)

Official website is: https://megatools.megous.com

If you'd like to donate, you can use:

PayPal: ahyechouchahquaachai@xff.cz
Patreon: https://www.patreon.com/megous


Contributors
============

- Chris Tarazi <tarazichris@gmail.com>
- Tom Maneiro <tomman@tsdx.net.ve>
- bAndie91 <bandie9100@gmail.com>
- Alberto Garcia <berto@igalia.com>
- David Guillen Fandos <david@davidgf.net>
- Erik Nordstrøm <erik@nordstroem.no>
- Johnathan Jenkins <john@nixheads.co.uk>
- Kagami Hiiragi <kagami@genshiken.org>
- Matthew Schultz <mattsch@gmail.com>
- Michael Ledin <mledin89@gmail.com>
- Michael Ripley <zkxs00@gmail.com>
- Palmer Dabbelt <palmer@dabbelt.com>
- RealDolos <dolos@cock.li>
- Viktor (Icon) VAD <vad.viktor@gmail.com>
- cyrozap <cyrozap@gmail.com>
- nyuszika7h <nyuszika7h@gmail.com>
- protomouse <root@protomou.se>
- strupo <sheeit@users.noreply.github.com>
- wdlkmpx <wdlkmpx@gmail.com>
- dal1a <waveh@trimsj.com>


Support and bug reports
=======================

If you think you've found bug in megatools, send a report including enough
information for recreating the issue to: megous@megous.com

IMPORTANT: I use personal e-mail server for my e-mail communication. I respond
to most e-mails, so if you are not getting a response within a few days, and
you're a user of a more aggressive e-mail provider (Outlook, iCloud, ...) it may
be because your provider rejected my e-mail as SPAM and did not deliver it at
all (not even to your SPAM folder), despite it being response to your e-mail and
my server using industry's best practices. You can either contact your e-mail
provider so that they fix their issue with rejecting legitimate responses to
your mail, or use different e-mail provider or contact channel to contact me.
(github issues) 


License
=======

Megatools are licensed under GPLv2 with OpenSSL exemption, see LICENSE
file for details.

This product includes software developed by the OpenSSL Project for use
in the OpenSSL Toolkit. (http://www.openssl.org/)


NEWS
megatools 1.10.2 - 2018-07-31
=============================

This is a bugfix release. This release was sponsored by donations to fix
large file downloads and improve download robustness after recent mega.nz
API change. Thank you!

Fixes:
- Actually fix connection dropping when downloading big files

Improvements:
- Add DNS resolution cache sharing accross all cURL handles
- Increase receive buffer size to 256kB

megatools 1.10.1 - 2018-07-27
=============================

This is a bugfix release + one feature addition by del1a.

New features:
- Allow to interactively choose which files to download from a public folder via
  megadl --choose-files

Fixes:
- Connection dropping when downloading big files
- Compatibility with older libcurl releases
- Fix occasional crashes when starting transfer worker threads
- Fix clang compatibility

megatools 1.10.0 - 2018-07-22
=============================

This release contains new features and optimizaions.

New features/optimizations:
- Chunked upload using up to 16 concurrent connections. Remember, more is not
  always better.
- Improved upload stability. When mega drops or hangs the data connection,
  megatools will simply restart the upload of the chunk of data that got lost.
- 5x AES enryption/decryption speed increase for uploads and downloads. Now
  it's possible to max out a gigabit connection on a cheap Intel based VPS.
- Improved progress reporting wih a summary of the average speed at the end.
- You can use --debug http to see all HTTP requests and connections, for
  debugging connection isues.
- Error 509 is reported with explanation.
- Implemented automatic resume for interrupted downloads. Megatools now
  writes data to a temporary file and renames it to the target name, when the
  download is fully completed. This way you'll notice that download was
  interrupted by CTRL+C and megatools will not leave half finished work behind.
- Add contributed bash auto-completion file by albaldi #368
- Add support for authenticated downloads of exported files using megadl #298
  (So that your account download quota is used instead of public quota.)
- Add Upload section and CreatePreviews setting into rc file. Add --enable-previews
  option.
- Add support for megals --print0|-0

Fixes:
- Upload hangs at 100% #366 #360 #365
- Don't send tetrminal escape sequences, when redirecting stdout.
- Fix comments syntax in megarc man page #359
- Use glib's variants of PRIu64, fixes #328
- megadl doesn't use account session when downloading folder links #304
- Skip symlinked folders when uploading in megacopy #262
- Turn on TCP keepalive probes in CURL #271
- Fix socks proxy support in megareg (it was not enabled correctly) #287
- Fix OpenSSL 1.1 compatibility #263
- Fix compiling against libressl
- Fix b64_aes128_cbc_encrypt_str string length handling #242
- Minor docummentation updates/fixes

Credits:
- Thanks to ERap320, megatools is also available on Chocolatey (see gihub issue
  #347) (If you want to verify origin of the binaries, because they are not
  distributed in the officialy signed zip file)
- Thanks to christarazi, megatools supports downloading specific sub-nodes of
  a public folder. (#254)

Thanks also go to all other contributors for improving documentation, reporting
bugs and testing.

megatools 1.9.98 - 2016-11-03
=============================

Bugfix release with some UI imporvements.

New features:
- Support upload/download speed limit settings
- Support socks proxy
- Improved progress reporting
- Support for OpenSSL 1.1.x

Removals:
- Remove undocummented --abort-on-error option. Tools now always
  report errors through exit status.
- Remove libmega.so public library support and a lot of unused code
  that was planned to be used for 2.0
- Remove megamv (it was never implemented and confused users)
- Remove megafs (it was just an experiment and confused users)

Cleanups:
- Cleanup build system a bit
- Cleanup CLI option handling, improved --help output

Fixes:
- Enable automatic decompression (CURLOPT_ACCEPT_ENCODING) (by protomouse)
  (This finally fixes the problem with HTTP compression.)
- Exit status from all the tools is now correctly reported
- Fix syncing of symlinked files
- Support very long passwords in the password prompt (up to 1024 chars)

megatools 1.9.97 - 2016-02-02
=============================

Bugfix release.

Fixes:
- Mega started compressing HTTP responses to API calls. Megatools now
  uses libcurl to handle API requests, so HTTP compression is now 
  supproted.

megatools 1.9.96 - 2016-01-02
=============================

Bugfix release.

Fixes:
- Refer to mega.nz and use mega.nz links instead of deprecated mega.co.nz
- Fix various build issues

megatools 1.9.95 - 2015-04-15
=============================

Bugfix and optimization release.

Fixes:
- Improve compatibility wit older curl (whatever version is in centos 6.5)
- Fix regression in megadl: fails to download exported folders #90
- Remove megamv as it is not and will not be implemented in 1.9 branch
- Slow down progress reporting on windows to once per second #102
- Correct the default values in mega_session_load()

Improvements:
- Rename megasync to megacopy to avoid conflict with official mega sync client
- Optimize session loading - issue #100
- Optimize session loading - further optimization - issue #100
- There's no need to use dbus-launcher when running megatools outside running
  desktop to avoid a warning message.

megatools 1.9.94 - 2015-01-02
=============================

Bugfix and optimization release.

Fixes:
- Force use of IPv4 addresses
- Fix path sanitization so that trailing slashes are ignored
- Document config file oddities in megarc manpage

lib/tools.c
/*
 *  megatools - Mega.nz client library and tools
 *  Copyright (C) 2013  Ondřej Jirman <megous@megous.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <curl/curl.h>
#include <openssl/err.h>
#include <openssl/ssl.h>
#include <locale.h>

#include "config.h"
#include "tools.h"

#ifdef G_OS_WIN32
#include <windows.h>
#else
#include <signal.h>
#include <termios.h>
#include <unistd.h>
#endif

#ifdef G_OS_WIN32
#define MEGA_RC_FILENAME "mega.ini"
#else
#define MEGA_RC_FILENAME ".megarc"
#endif

#define BOOLEAN_UNSET_BUT_TRUE 2

static GOptionContext *opt_context;
static gchar *opt_username;
static gchar *opt_password;
static gchar *opt_config;
static gboolean opt_reload_files;
static gboolean opt_version;
static gboolean opt_no_config;
static gboolean opt_no_ask_password;
static gint opt_speed_limit = -1; /* -1 means limit not set */
static gchar *opt_proxy;

static gchar *proxy;
static gint upload_speed_limit;
static gint download_seed_limit;
static gint transfer_worker_count = 5;
static gint cache_timout = 10 * 60;
static gboolean opt_enable_previews = BOOLEAN_UNSET_BUT_TRUE;
static gboolean opt_disable_resume;

static gboolean opt_debug_callback(const gchar *option_name, const gchar *value, gpointer data, GError **error)
{
	if (value) {
		gchar **opts = g_strsplit(value, ",", 0);
		gchar **opt = opts;

		while (*opt) {
			if (g_ascii_strcasecmp(*opt, "api") == 0)
				mega_debug |= MEGA_DEBUG_API;
			else if (g_ascii_strcasecmp(*opt, "fs") == 0)
				mega_debug |= MEGA_DEBUG_FS;
			else if (g_ascii_strcasecmp(*opt, "cache") == 0)
				mega_debug |= MEGA_DEBUG_CACHE;
			else if (g_ascii_strcasecmp(*opt, "http") == 0)
				mega_debug |= MEGA_DEBUG_HTTP;
			else if (g_ascii_strcasecmp(*opt, "tman") == 0)
				mega_debug |= MEGA_DEBUG_TMAN;

			opt++;
		}

		g_strfreev(opts);
	} else {
		mega_debug = MEGA_DEBUG_API;
	}

	return TRUE;
}

static GOptionEntry basic_options[] = {
	{ "config", '\0',
                0, G_OPTION_ARG_FILENAME, &opt_config,
                "Load configuration from a file", "PATH" },
	{ "ignore-config-file", '\0',
                0, G_OPTION_ARG_NONE, &opt_no_config,
                "Disable loading " MEGA_RC_FILENAME, NULL },
	{ "debug", '\0',
                G_OPTION_FLAG_OPTIONAL_ARG, G_OPTION_ARG_CALLBACK, opt_debug_callback,
                "Enable debugging output", "OPTS" },
	{ "version", '\0',
                0, G_OPTION_ARG_NONE, &opt_version,
                "Show version information", NULL },
	{ NULL }
};

static GOptionEntry upload_options[] = {
        { "enable-previews", '\0',
                0, G_OPTION_ARG_NONE, &opt_enable_previews,
                "Generate previews when uploading file", NULL },
        { "disable-previews", '\0',
                G_OPTION_FLAG_REVERSE, G_OPTION_ARG_NONE, &opt_enable_previews,
                "Never generate previews when uploading file", NULL },
        { NULL }
};

static GOptionEntry download_options[] = {
        { "disable-resume", '\0',
                0, G_OPTION_ARG_NONE, &opt_disable_resume,
                "Disable resume when downloading file", NULL },
        { NULL }
};

static GOptionEntry auth_options[] = {
	{ "username", 'u',
                0, G_OPTION_ARG_STRING, &opt_username,
                "Account username (email)", "USERNAME" },
	{ "password", 'p',
                0, G_OPTION_ARG_STRING, &opt_password,
                "Account password", "PASSWORD" },
	{ "no-ask-password", '\0',
                0, G_OPTION_ARG_NONE, &opt_no_ask_password,
                "Never ask interactively for a password", NULL },
	{ "reload", '\0',
                0, G_OPTION_ARG_NONE, &opt_reload_files,
                "Reload filesystem cache", NULL },
	{ NULL }
};

static GOptionEntry network_options[] = {
	{ "limit-speed", '\0',
                0, G_OPTION_ARG_INT, &opt_speed_limit,
                "Limit transfer speed (KiB/s)", "SPEED" },
	{ "proxy", '\0',
                0, G_OPTION_ARG_STRING, &opt_proxy,
                "Proxy setup string", "PROXY" },
	{ NULL }
};

#if OPENSSL_VERSION_NUMBER >= 0x10100004L
static void init_openssl_locking()
{
	// OpenSSL >= 1.1.0-pre4 doesn't require specific callback setup
}
#else
#if GLIB_CHECK_VERSION(2, 32, 0)

static GMutex *openssl_mutexes = NULL;

static void openssl_locking_callback(int mode, int type, const char *file, int line)
{
	if (mode & CRYPTO_LOCK)
		g_mutex_lock(openssl_mutexes + type);
	else
		g_mutex_unlock(openssl_mutexes + type);
}

static unsigned long openssl_thread_id_callback()
{
	unsigned long ret;
	ret = (unsigned long)g_thread_self();
	return ret;
}

static void init_openssl_locking()
{
	gint i;

	// initialize OpenSSL locking for multi-threaded operation
	openssl_mutexes = g_new(GMutex, CRYPTO_num_locks());
	for (i = 0; i < CRYPTO_num_locks(); i++)
		g_mutex_init(openssl_mutexes + i);

	SSL_library_init();
	CRYPTO_set_id_callback(openssl_thread_id_callback);
	CRYPTO_set_locking_callback(openssl_locking_callback);
}

#else

static GMutex **openssl_mutexes = NULL;

static void openssl_locking_callback(int mode, int type, const char *file, int line)
{
	if (mode & CRYPTO_LOCK)
		g_mutex_lock(openssl_mutexes[type]);
	else
		g_mutex_unlock(openssl_mutexes[type]);
}

static unsigned long openssl_thread_id_callback()
{
	unsigned long ret;
	ret = (unsigned long)g_thread_self();
	return ret;
}

static void init_openssl_locking()
{
	gint i;

	// initialize OpenSSL locking for multi-threaded operation
	openssl_mutexes = g_new(GMutex *, CRYPTO_num_locks());
	for (i = 0; i < CRYPTO_num_locks(); i++)
		openssl_mutexes[i] = g_mutex_new();

	SSL_library_init();
	CRYPTO_set_id_callback(openssl_thread_id_callback);
	CRYPTO_set_locking_callback(openssl_locking_callback);
}

#endif
#endif

static void init(void)
{
#if !GLIB_CHECK_VERSION(2, 32, 0)
	if (!g_thread_supported())
		g_thread_init(NULL);
#endif

	setlocale(LC_ALL, "");

#if !GLIB_CHECK_VERSION(2, 36, 0)
	g_type_init();
#endif

	g_setenv("GSETTINGS_BACKEND", "memory", TRUE);

#ifndef G_OS_WIN32
	signal(SIGPIPE, SIG_IGN);
#endif

	init_openssl_locking();
}

gboolean tool_is_stdout_tty(void)
{
#ifdef G_OS_WIN32
	return FALSE;
#else
	return isatty(1);
#endif
}

#define PROGRESS_FREQUENCY ((gint64)1000000)

void tool_show_progress(const gchar *file, const struct mega_status_data *data)
{
	static gint64 last_update = 0;
	static guint64 last_bytes = 0;
	static gint64 transfer_start = -1;

	gint64 timenow = g_get_monotonic_time();
	gint64 now_done = 0;
	gc_free gchar *done_str = NULL;
	gc_free gchar *total_str = NULL;
	gc_free gchar *rate_str = NULL;

	if (data->progress.total == 0)
		return;

	// start of the new transfer, initialize progress reporting
	if (data->progress.done == -1) {
		transfer_start = last_update = timenow;
		last_bytes = 0;
	} else if (transfer_start < 0)
		return;
	else if (data->progress.done == -2)
		now_done = data->progress.total;
	else if (last_update && last_update + PROGRESS_FREQUENCY > timenow)
		return;
	else
		now_done = data->progress.done;

	gint64 time_span = timenow - last_update;
	gint64 size_diff = now_done - last_bytes;
	gdouble rate, percentage;

	if (data->progress.done == -2) {
		// final summary
		rate = (gdouble)data->progress.total * 1e6 / (timenow - transfer_start);
		percentage = 100;
		transfer_start = -1;

		total_str = g_format_size_full(data->progress.total, G_FORMAT_SIZE_IEC_UNITS);
		rate_str = g_format_size_full(rate, G_FORMAT_SIZE_IEC_UNITS);

		if (tool_is_stdout_tty()) {
			g_print(ESC_WHITE "%s" ESC_NORMAL ": "
				ESC_YELLOW "%.2f%%" ESC_NORMAL " - "
				"done " ESC_GREEN "%s" ESC_NORMAL
				" (avg. %s/s)" ESC_CLREOL "\r",
				file, percentage, total_str, rate_str);
		} else {
			g_print("%s: %.2f%% - done %s (avg. %s/s)\n", file, percentage, total_str, rate_str);
		}
	} else if (time_span == 0) {
		// just started
		percentage = 0;

		done_str = g_format_size_full(now_done, G_FORMAT_SIZE_IEC_UNITS | G_FORMAT_SIZE_LONG_FORMAT);
		total_str = g_format_size_full(data->progress.total, G_FORMAT_SIZE_IEC_UNITS);

		if (tool_is_stdout_tty()) {
			g_print(ESC_WHITE "%s" ESC_NORMAL ": "
				ESC_YELLOW "%.2f%%" ESC_NORMAL " - "
				ESC_GREEN "%s" ESC_BLUE " of %s" ESC_NORMAL
				ESC_CLREOL "\r",
				file, percentage, done_str, total_str);
		} else {
			g_print("%s: %.2f%% - %s of %s\n", file, percentage, done_str, total_str);
		}
	} else {
		// regular update
		rate = (gdouble)size_diff * 1e6 / time_span;
		percentage = (gdouble)now_done / data->progress.total * 100;

		done_str = g_format_size_full(now_done, G_FORMAT_SIZE_IEC_UNITS | G_FORMAT_SIZE_LONG_FORMAT);
		total_str = g_format_size_full(data->progress.total, G_FORMAT_SIZE_IEC_UNITS);
		rate_str = g_format_size_full(rate, G_FORMAT_SIZE_IEC_UNITS);

		if (tool_is_stdout_tty()) {
			g_print(ESC_WHITE "%s" ESC_NORMAL ": "
				ESC_YELLOW "%.2f%%" ESC_NORMAL " - "
				ESC_GREEN "%s" ESC_BLUE " of %s" ESC_NORMAL
				" (%s/s)" ESC_CLREOL "\r",
				file, percentage, done_str, total_str, rate_str);
		} else {
			g_print("%s: %.2f%% - %s of %s (%s/s)\n", file, percentage, done_str, total_str, rate_str);
		}
	}

	last_update = timenow;
	last_bytes = now_done;
}

gchar* tool_prompt_input(void)
{
	gc_string_free GString* str = g_string_sized_new(1024);
	gchar* ret = NULL;

	while (TRUE) {
		char buf[256];

		if (fgets(buf, sizeof buf, stdin) == NULL)
			return NULL;

		int len = strlen(buf);
		gboolean has_eol = buf[len - 1] == '\n';
		if (len > 0)
			g_string_append_len(str, buf, has_eol ? len - 1 : len);

		if (has_eol || feof(stdin))
			break;
	}

	ret = g_string_free(str, FALSE);
	str = NULL;
	return ret;
}

static gchar *input_password(void)
{
	gint tries = 3;
	gchar buf[1025];
	gchar *password = NULL;

#ifdef G_OS_WIN32
	HANDLE hStdin = GetStdHandle(STD_INPUT_HANDLE);
	DWORD mode = 0;
	GetConsoleMode(hStdin, &mode);
	SetConsoleMode(hStdin, mode & (~ENABLE_ECHO_INPUT));
#else
	struct termios oldt;
	tcgetattr(STDIN_FILENO, &oldt);
	struct termios newt = oldt;
	newt.c_lflag &= ~ECHO;
	tcsetattr(STDIN_FILENO, TCSANOW, &newt);
#endif

again:
	g_print("Enter password for (%s): ", opt_username);
	if (fgets(buf, 1024, stdin)) {
		if (strlen(buf) > 1) {
			password = g_strndup(buf, strcspn(buf, "\r\n"));
		} else {
			if (--tries > 0) {
				g_print("\n");
				goto again;
			}

			g_print("\nYou need to provide non-empty password!\n");
			exit(1);
		}
	} else {
		g_printerr("\nERROR: Can't read password from the input!\n");
		exit(1);
	}

#ifdef G_OS_WIN32
	SetConsoleMode(hStdin, mode);
#else
	tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
#endif

	g_print("\nGood, signing in...\n");

	return password;
}

static void print_version(void)
{
	if (opt_version) {
		g_print("megatools " VERSION " - command line tools for Mega.nz\n\n");
		g_print("Written by Ondrej Jirman <megous@megous.com>, 2013-2018\n");
		g_print("Go to http://megatools.megous.com for more information\n");
		exit(0);
	}
}

gchar *tool_convert_filename(const gchar *path, gboolean local)
{
	gchar *locale_path;

#ifdef G_OS_WIN32
	locale_path = g_locale_to_utf8(path, -1, NULL, NULL, NULL);
#else
	if (local)
		locale_path = g_strdup(path);
	else
		locale_path = g_locale_to_utf8(path, -1, NULL, NULL, NULL);
#endif

	if (locale_path == NULL) {
		g_printerr(
			"ERROR: Invalid filename locale, can't convert file names specified on the command line to UTF-8.\n");
		exit(1);
	}

	return locale_path;
}

void tool_init(gint *ac, gchar ***av, const gchar *tool_name, GOptionEntry *tool_entries, ToolInitFlags flags)
{
	GError *local_err = NULL;

	init();

	opt_context = g_option_context_new(tool_name);
	if (tool_entries)
		g_option_context_add_main_entries(opt_context, tool_entries, NULL);

	if (flags & TOOL_INIT_UPLOAD_OPTS)
		g_option_context_add_main_entries(opt_context, upload_options, NULL);

	if (flags & TOOL_INIT_DOWNLOAD_OPTS)
		g_option_context_add_main_entries(opt_context, download_options, NULL);

	if (flags & (TOOL_INIT_AUTH | TOOL_INIT_AUTH_OPTIONAL))
		g_option_context_add_main_entries(opt_context, auth_options, NULL);

	g_option_context_add_main_entries(opt_context, network_options, NULL);
	g_option_context_add_main_entries(opt_context, basic_options, NULL);

	if (!g_option_context_parse(opt_context, ac, av, &local_err)) {
		g_printerr("ERROR: Option parsing failed: %s\n", local_err->message);
		g_clear_error(&local_err);
		exit(1);
	}

	print_version();

	if (!opt_no_config || opt_config) {
		gboolean status = TRUE;
		gc_key_file_unref GKeyFile *kf = g_key_file_new();

		if (opt_config) {
			if (!g_key_file_load_from_file(kf, opt_config, 0, &local_err)) {
				g_printerr("ERROR: Failed to open config file: %s: %s\n", opt_config,
					   local_err->message);
				g_clear_error(&local_err);
				exit(1);
			}
		} else {
			status = g_key_file_load_from_file(kf, MEGA_RC_FILENAME, 0, NULL);
			if (!status) {
				gc_free gchar *tmp = g_build_filename(g_get_home_dir(), MEGA_RC_FILENAME, NULL);
				status = g_key_file_load_from_file(kf, tmp, 0, NULL);
			}
		}

		if (status) {
			// load username/password from ini file
			if (!opt_username)
				opt_username = g_key_file_get_string(kf, "Login", "Username", NULL);

			if (!opt_password)
				opt_password = g_key_file_get_string(kf, "Login", "Password", NULL);

			gint to = g_key_file_get_integer(kf, "Cache", "Timeout", &local_err);
			if (local_err == NULL)
				cache_timout = to;
			else
				g_clear_error(&local_err);

			// Load speed limits from settings file
			if (g_key_file_has_key(kf, "Network", "SpeedLimit", NULL)) {
				download_seed_limit = upload_speed_limit =
					g_key_file_get_integer(kf, "Network", "SpeedLimit", &local_err);
				if (local_err) {
					g_printerr("WARNING: Invalid speed limit set in the config file: %s\n",
						   local_err->message);
					g_clear_error(&local_err);
				}
			}

			if (g_key_file_has_key(kf, "Network", "UploadSpeedLimit", NULL)) {
				upload_speed_limit =
					g_key_file_get_integer(kf, "Network", "UploadSpeedLimit", &local_err);
				if (local_err) {
					g_printerr("WARNING: Invalid upload speed limit set in the config file: %s\n",
						   local_err->message);
					g_clear_error(&local_err);
				}
			}

			if (g_key_file_has_key(kf, "Network", "DownloadSpeedLimit", NULL)) {
				download_seed_limit =
					g_key_file_get_integer(kf, "Network", "DownloadSpeedLimit", &local_err);
				if (local_err) {
					g_printerr("WARNING: Invalid download speed limit set in the config file: %s\n",
						   local_err->message);
					g_clear_error(&local_err);
				}
			}

			if (g_key_file_has_key(kf, "Network", "ParallelTransfers", NULL)) {
				transfer_worker_count =
					g_key_file_get_integer(kf, "Network", "ParallelTransfers", &local_err);
				if (local_err) {
					g_printerr(
						"WARNING: Invalid number of parallel transfers set in the config file: %s\n",
						local_err->message);
					g_clear_error(&local_err);
				}

				if (transfer_worker_count < 1 || transfer_worker_count > 16) {
					transfer_worker_count = CLAMP(transfer_worker_count, 1, 16);
					g_printerr(
						"WARNING: Invalid number of parallel transfers set in the config file, limited to %d\n",
						transfer_worker_count);
				}
			}

			proxy = g_key_file_get_string(kf, "Network", "Proxy", NULL);

			if (opt_enable_previews == BOOLEAN_UNSET_BUT_TRUE) {
				gboolean enable = g_key_file_get_boolean(kf, "Upload", "CreatePreviews", &local_err);
				if (local_err == NULL)
					opt_enable_previews = enable;
				else
					g_clear_error(&local_err);
			}
		}
	}

	if (opt_speed_limit >= 0) {
		upload_speed_limit = opt_speed_limit;
		download_seed_limit = opt_speed_limit;
	}

	if (opt_proxy) {
		if (!strcmp(opt_proxy, "none"))
			proxy = NULL;
		else
			proxy = opt_proxy;
	}

	if (!(flags & TOOL_INIT_AUTH))
		return;

	if (!opt_username) {
		g_printerr("ERROR: You must specify your mega.nz username (email)\n");
		exit(1);
	}

	if (!opt_password && opt_no_ask_password) {
		g_printerr("ERROR: You must specify your mega.nz password\n");
		exit(1);
	}

	if (!opt_password)
		opt_password = input_password();
}

struct mega_session *tool_start_session(ToolSessionFlags flags)
{
	GError *local_err = NULL;
	gchar *sid = NULL;
	gboolean loaded = FALSE;

	struct mega_session *s = mega_session_new();

	mega_session_set_speed(s, upload_speed_limit, download_seed_limit);
	mega_session_set_workers(s, transfer_worker_count);

	if (proxy)
		mega_session_set_proxy(s, proxy);

	mega_session_enable_previews(s, TRUE);

	if (!(flags & TOOL_SESSION_OPEN))
		return s;

	// allow unatuhenticated session
	if (!opt_password || !opt_username) {
		if (flags & TOOL_SESSION_AUTH_OPTIONAL)
			return s;

		g_printerr("ERROR: Authentication is required\n");
		goto err;
	}

	// try to load cached session data (they are valid for 10 minutes since last
	// user_get or refresh)
	if (!mega_session_load(s, opt_username, opt_password, cache_timout, &sid, &local_err)) {
		g_clear_error(&local_err);

		if (!mega_session_open(s, opt_username, opt_password, sid, &local_err)) {
			g_printerr("ERROR: Can't login to mega.nz: %s\n", local_err->message);
			goto err;
		}

		if (!(flags & TOOL_SESSION_AUTH_ONLY)) {
			if (!mega_session_refresh(s, &local_err)) {
				g_printerr("ERROR: Can't read filesystem info from mega.nz: %s\n", local_err->message);
				goto err;
			}

			loaded = TRUE;
		}

		mega_session_save(s, NULL);
	}

	if (!(flags & TOOL_SESSION_AUTH_ONLY) && opt_reload_files && !loaded) {
		if (!mega_session_refresh(s, &local_err)) {
			g_printerr("ERROR: Can't read filesystem info from mega.nz: %s\n", local_err->message);
			goto err;
		}

		mega_session_save(s, NULL);
	}

	mega_session_enable_previews(s, !!opt_enable_previews);
	mega_session_set_resume(s, !opt_disable_resume);

	g_free(sid);
	return s;

err:
	mega_session_free(s);
	g_clear_error(&local_err);
	g_free(sid);
	return NULL;
}

void tool_fini(struct mega_session *s)
{
	if (s)
		mega_session_free(s);

	mega_cleanup();

	g_option_context_free(opt_context);
	curl_global_cleanup();
	CRYPTO_cleanup_all_ex_data();
	ERR_free_strings();

#if OPENSSL_VERSION_NUMBER < 0x10100004L
	CRYPTO_set_id_callback(NULL);
	CRYPTO_set_locking_callback(NULL);
#endif
}
